src/net.cpp

```diff
@@ -718,6 +720,170 @@ CNetMessage V1TransportDeserializer::GetMessage(const CMessageHeader::MessageSta
     return msg;
 }
 
+int V2TransportDeserializer::Read(const char* pch, unsigned int bytes)
+{
+    if (!m_in_data) {
+        // copy data to temporary parsing buffer
+        const unsigned int remaining = CHACHA20_POLY1305_AEAD_AAD_LEN - m_hdr_pos;
+        const unsigned int copy_bytes = std::min(remaining, bytes);
+
+        memcpy(&vRecv[m_hdr_pos], pch, copy_bytes);
+        m_hdr_pos += copy_bytes;
+
+        // if AAD incomplete, exit
+        if (m_hdr_pos < CHACHA20_POLY1305_AEAD_AAD_LEN) {
+            return copy_bytes;
+        }
+
+        // we got the AAD bytes at this point (3 bytes encrypted packet length)
+        // we keep the sequence numbers unchanged at this point. Once the message is authenticated and decrypted, we increase the sequence numbers (or the aad_pos)
+        if (!m_aead->GetLength(&m_message_size, m_aad_seqnr, m_aad_pos, (const uint8_t*)vRecv.data())) {
+            return -1;
+        }
+
+        // check and unset rekey bit
+        // the counterparty can signal a post-this-message rekey by setting the
+        // most significant bit in the (unencrypted) length
+        m_rekey_flag = static_cast<bool>(m_message_size & (1U << 23));
+        if (m_rekey_flag) {
+            LogPrint(BCLog::NET, "Rekey flag detected\n");
+            m_message_size &= ~(1U << 23);
+        }
+
+        // reject messages larger than MAX_SIZE
+        if (m_message_size > MAX_SIZE) {
+            return -1;
```

Comment:
I agree that this would be a nice cleanup. But it should also be done for the V1 deserializer. I think we should do that after this PR since it contains lines not changes otherwise.